apm-agent-plugins/apm-jms-plugin/src/main/java/co/elastic/apm/agent/jms/JmsInstrumentationHelperImpl.java

```diff
@@ -110,20 +137,74 @@ public void onMessage(Message message) {
+                    destinationName = ((Topic) destination).getTopicName();
+                }
+            } catch (JMSException e) {
+                logger.error("Failed to obtain destination name", e);
+            }
+        }
+        return destinationName;
+    }
+
+    private boolean isTempDestination(Destination destination, @Nullable String extractedDestinationName) {
+        return destination instanceof TemporaryQueue || destination instanceof TemporaryTopic ||
+            (extractedDestinationName != null && extractedDestinationName.startsWith(TMP_PREFIX));
+    }
+
+    @Override
+    public boolean ignoreDestination(@Nullable String destinationName) {
+        return WildcardMatcher.isAnyMatch(messagingConfiguration.getIgnoreMessageQueues(), destinationName);
+    }
+
+    @Override
+    public void addDestinationDetails(@Nullable Message message, Destination destination, String destinationName,
+                                      AbstractSpan span) {
+        if (destination instanceof Queue) {
+            span.appendToName("queue ").appendToName(destinationName)
+                .getContext().getMessage().withQueue(destinationName);
+        } else if (destination instanceof Topic) {
+            span.appendToName("topic ").appendToName(destinationName)
+                .getContext().getMessage().withTopic(destinationName);
+        }
+    }
+
     @Override
-    public void addDestinationDetails(Destination destination, AbstractSpan span) {
+    public void addMessageDetails(@Nullable Message message, AbstractSpan span) {
+        if (message == null) {
+            return;
+        }
         try {
-            if (destination instanceof Queue) {
-                String queueName = ((Queue) destination).getQueueName();
-                span.appendToName("queue ").appendToName(queueName)
-                    .getContext().getMessage().withQueue(queueName);
-            } else if (destination instanceof Topic) {
-                String topicName = ((Topic) destination).getTopicName();
-                span.appendToName("topic ").appendToName(topicName)
-                    .getContext().getMessage().withTopic(topicName);
+            co.elastic.apm.agent.impl.context.Message messageContext = span.getContext().getMessage();
+
+            // Currently only capturing body of TextMessages. The javax.jms.Message#getBody() API is since 2.0, so,
+            // if we are supporting JMS 1.1, it makes no sense to rely on isAssignableFrom.
+            if (coreConfiguration.getCaptureBody() != CoreConfiguration.EventType.OFF && message instanceof TextMessage) {
+                messageContext.withBody(((TextMessage) message).getText());
+            }
+
+            // Addition of non-String headers/properties will cause String instance allocations
+            if (coreConfiguration.isCaptureHeaders()) {
+                messageContext.addHeader(JMS_MESSAGE_ID_HEADER, message.getJMSMessageID());
```

Comment:
This only gets some headers, why not get all via `getPropertyNames`/`getObjectProperty`?

Reply:
In JMS, headers are predefined and properties are custom. 
For headers- we don't have an API to get by name, only through dedicated APIs. So I chose the three I thought would be most interesting. Once another is requested, it is now super-easy to add.

As for properties- the `javax.jms.Message` only offers `getXXXProperty()` APIs, where `XXX` is an explicit type... So, the only way to get all properties is to iterate over the `Enumeration` returned from `getPropertyNames()` and try all types one-by-one....
I decided it is too much for now and we can reconsider if we see demand.

Reply:
Ohh, you're right! I completely overlooked it- the `getObjectProperty` is actually what I need. Thanks!!